### PR TITLE
Add support for time limited credentials auth

### PR DESCRIPTION
**Pull request by G-Kemp101**
Add support for time limited credential authentication as according to https://datatracker.ietf.org/doc/html/draft-uberti-behave-turn-rest-00.

Also refactors the naming of decode/encode so it does not overwrite the `bytes` base implementation

### CODE REVIEW

File: jostedal/stun/authentication.py [hunk 1]
```diff
@@ -35,18 +37,46 @@ def __init__(self, realm, users):
...
             if not key:
                 password = credentials.get("password")
                 if not password:
                     logger.warning("Invalid credentials for %s", username)
                     continue
-            self.hmac_keys[username] = ha1(username, self.realm, password)
+
+                self.hmac_keys[username] = ha1(username, self.realm, password)
+            else:
+                self.credential_type = "shared_secret"
```
Comment (frederik-vestre-pexip) [inline jostedal/stun/authentication.py:51]: Is it a posibility that users contains multiple entries with different types (i.e some "namepsw", and some "shared_secret"). How would this work then?

File: jostedal/utils.py [hunk 2]
```diff
@@ -10,3 +15,17 @@ def ha1(username, realm, password):
...
     ).digest()
+
+
+def secret_key(username, realm, password):
+    """
+    Generate passwords for Coturn REST api. See the "TURN REST API" section in
+    README.turnserver in the coturn source code for algorithm details.
+    """
+    # Using sha1 because it's required by the TURN server API
+    print(type(password))
```
Comment (frederik-vestre-pexip) [inline jostedal/utils.py]: do we want to use print in test code. Do we have any more specific log functionality we can use or is this temporary code that could be removed?

**Comment by frederik-vestre-pexip** [comment]
Not sure if i can review the from_str / from_buffer renames, they are probably nice, but i don't know enough about the code to judge them so i have not looked at any of those changes.

File: jostedal/stun/authentication.py (hunk 1 above)
Comment (G-Kemp101) [reply jostedal/stun/authentication.py:51]: I don't think the code would have ever worked like that, as the message integrity https://github.com/pexip/jostedal/blob/master/jostedal/stun/authentication.py#L56 would always pick the first element in the dictionary anyway

File: jostedal/utils.py (hunk 2 above)
Comment (G-Kemp101) [reply jostedal/utils.py]: No sorry, left over from debugging

File: jostedal/stun/authentication.py (hunk 1 above)
Comment (frederik-vestre-pexip) [reply jostedal/stun/authentication.py:51]: ok, i guess that is fair. Not sure what the point of the loop is in any case then (except printing a warning), but i guess it is not important.

**Comment by johnbassett** [comment]
👍

**Comment by frederik-vestre-pexip** [comment]
:+1:

**Pull request merged**